staging/src/k8s.io/controller-manager/pkg/controllers/lease/controller.go

```diff
@@ -170,12 +180,16 @@ func (c *controller) ensureLease() (*coordinationv1.Lease, bool, error) {
 // call this once you're sure the lease has been created
 func (c *controller) retryUpdateLease(base *coordinationv1.Lease) error {
 	for i := 0; i < maxUpdateRetries; i++ {
-		lease, err := c.leaseClient.Update(context.TODO(), c.newLease(base), metav1.UpdateOptions{})
-		if err == nil {
-			c.latestLease = lease
-			return nil
+		leaseToUpdate, blockUpdate, err := c.newLease(base)
+		if !blockUpdate {
```

Comment:
Copy [comment](https://github.com/kubernetes/kubernetes/pull/95222#discussion_r500643719) from @caesarxuchao here:

> Can you use the opposite condition? i.e.,
```
if blockUpdate {
  klog.Error(...)
  continue
}
lease, err = c.leaseClient.Update(context.TODO(), leaseToUpdate, metav1.UpdateOptions{})
if err == nil {
  c.latestLease = lease
  return nil
}
klog.Error(...)
```
> Also move the onRepeatedHeartbeatFailure check to the beginning of the for loop to keep the semantics.

Reply:
I'm not sure if I understand it correctly, but I think moving `onRepeatedHeartbeatFailure` to the beginning changes the behavior, as the function may get invoked when the heartbeat operation failed just once: https://github.com/kubernetes/kubernetes/blob/77866160efce1ce4c3885d436b1cebaedc80c89c/pkg/kubelet/kubelet.go#L817-L818